403 causes panic

This was 100% a permission issue on my end but a 403 download code shouldn't cause a panic in the code.

```ps1
PS C:\Users\Clayton Burlison\Desktop\cpe-chef-repo> C:\ProgramData\gorilla\bin\gorilla.exe -v
Retrieving manifest: unity_prd
Manifest Url: https://example.com/gorilla_repo/manifests/company_prd.yaml
panic: Unable to retrieve manifest:company_prdcompany_prd.yaml : Download status code: 403

goroutine 1 [running]:
log.Panic(0xc000067d40, 0x3, 0x3)
        /usr/local/go/src/log/log.go:326 +0xc7
github.com/1dustindavis/gorilla/pkg/gorillalog.Error(0xc000067d40, 0x3, 0x3)
        /ext-go/1/src/github.com/1dustindavis/gorilla/pkg/gorillalog/gorillalog.go:78 +0x5f
github.com/1dustindavis/gorilla/pkg/manifest.Get(0xc000067f68, 0x2, 0x2)
        /ext-go/1/src/github.com/1dustindavis/gorilla/pkg/manifest/manifest.go:68 +0x7b4
main.main()
        /ext-go/1/src/github.com/1dustindavis/gorilla/cmd/gorilla/main.go:25 +0xd0
```

Whatever does that double manifest name to come up should also be fixed. I saw that as well.

Thanks,
Erik Gomez
________________________________
From: Clayton Burlison <notifications@github.com>
Sent: Tuesday, March 12, 2019 11:54:35 AM
To: 1dustindavis/gorilla
Cc: Subscribed
Subject: [1dustindavis/gorilla] 403 causes panic (#65)


This was 100% a permission issue on my end but a 403 download code shouldn't cause a panic in the code.

PS C:\Users\Clayton Burlison\Desktop\cpe-chef-repo> C:\ProgramData\gorilla\bin\gorilla.exe -v
Retrieving manifest: unity_prd
Manifest Url: https://example.com/gorilla_repo/manifests/company_prd.yaml
panic: Unable to retrieve manifest:company_prdcompany_prd.yaml : Download status code: 403

goroutine 1 [running]:
log.Panic(0xc000067d40, 0x3, 0x3)
        /usr/local/go/src/log/log.go:326 +0xc7
github.com/1dustindavis/gorilla/pkg/gorillalog.Error(0xc000067d40, 0x3, 0x3)
        /ext-go/1/src/github.com/1dustindavis/gorilla/pkg/gorillalog/gorillalog.go:78 +0x5f
github.com/1dustindavis/gorilla/pkg/manifest.Get(0xc000067f68, 0x2, 0x2)
        /ext-go/1/src/github.com/1dustindavis/gorilla/pkg/manifest/manifest.go:68 +0x7b4
main.main()
        /ext-go/1/src/github.com/1dustindavis/gorilla/cmd/gorilla/main.go:25 +0xd0

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://github.com/1dustindavis/gorilla/issues/65>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AFa_GNLQMEOK6mmqRvXlVyhzW-ESE2bhks5vV9vLgaJpZM4brZTS>.

--- a/pkg/catalog/catalog.go
+++ b/pkg/catalog/catalog.go
@@ -1,6 +1,7 @@
 package catalog
 
 import (
+	"fmt"
 	"io/ioutil"
 	"os"
 	"path/filepath"
@@ -67,6 +68,14 @@ func Get(cfg config.Configuration) map[int]map[string]Item {
 		os.Exit(1)
 	}
 
+	// Setup to catch a potential failure
+	defer func() {
+		if r := recover(); r != nil {
+			fmt.Println(r)
+			os.Exit(1)
+		}
+	}()
+
 	// Loop through the catalogs and get each one in order
 	for _, catalog := range cfg.Catalogs {
 
@@ -77,7 +86,7 @@ func Get(cfg config.Configuration) map[int]map[string]Item {
 		gorillalog.Info("Catalog Url:", catalogURL)
 		err := downloadFile(cfg.CachePath, catalogURL)
 		if err != nil {
-			gorillalog.Error("Unable to retrieve catalog:", catalog, err)
+			gorillalog.Error("Unable to retrieve catalog: ", err)
 		}
 
 		// Open the catalog file
@@ -85,14 +94,14 @@ func Get(cfg config.Configuration) map[int]map[string]Item {
 		gorillalog.Debug("Catalog file path:", yamlPath)
 		yamlFile, err := ioutil.ReadFile(yamlPath)
 		if err != nil {
-			gorillalog.Error("Unable to open the catalog file:", yamlPath, err)
+			gorillalog.Error("Unable to open the catalog file: ", err)
 		}
 
 		// Parse the catalog
 		var catalogItems map[string]Item
 		err = yaml.Unmarshal(yamlFile, &catalogItems)
 		if err != nil {
-			gorillalog.Error("Unable to parse yaml catalog:", yamlPath, err)
+			gorillalog.Error("Unable to parse yaml catalog: ", err)
 		}
 
 		// Add the new parsed catalog items to the catalogMap

--- a/pkg/gorillalog/gorillalog.go
+++ b/pkg/gorillalog/gorillalog.go
@@ -25,7 +25,7 @@ func NewLog(cfg config.Configuration) {
 	// Setup a defer function to recover from a panic
 	defer func() {
 		if r := recover(); r != nil {
-			fmt.Println("Recovered from panic while creating the log file:", r)
+			fmt.Println(r)
 			os.Exit(1)
 		}
 	}()

--- a/pkg/installer/installer.go
+++ b/pkg/installer/installer.go
@@ -85,7 +85,7 @@ func installItem(item catalog.Item, itemURL, cachePath string) string {
 	if item.Installer.Type == "nupkg" {
 		gorillalog.Info("Installing nupkg for", item.DisplayName)
 		installCmd = commandNupkg
-		installArgs = []string{"install", absFile, "-f", "-y", "-r"}
+		installArgs = []string{"install", absFile, "-f", "-y"}
 	} else if item.Installer.Type == "msi" {
 		gorillalog.Info("Installing msi for", item.DisplayName)
 		installCmd = commandMsi
@@ -135,7 +135,7 @@ func uninstallItem(item catalog.Item, itemURL, cachePath string) string {
 	if item.Uninstaller.Type == "nupkg" {
 		gorillalog.Info("Installing nupkg for", item.DisplayName)
 		uninstallCmd = commandNupkg
-		uninstallArgs = []string{"uninstall", absFile, "-f", "-y", "-r"}
+		uninstallArgs = []string{"uninstall", absFile, "-f", "-y"}
 	} else if item.Uninstaller.Type == "msi" {
 		gorillalog.Info("Installing msi for", item.DisplayName)
 		uninstallCmd = commandMsi

--- a/pkg/installer/installer.go
+++ b/pkg/installer/installer.go
@@ -180,13 +180,16 @@ func Install(item catalog.Item, installerType, urlPackages, cachePath string) st
 		return "Item not needed"
 	}
 
-	// Compile the item's URL
-	itemURL := urlPackages + item.Uninstaller.Location
-
 	// Install or uninstall the item
 	if installerType == "install" || installerType == "update" {
+		// Compile the item's URL
+		itemURL := urlPackages + item.Installer.Location
+		// Run the installer
 		installItem(item, itemURL, cachePath)
 	} else if installerType == "uninstall" {
+		// Compile the item's URL
+		itemURL := urlPackages + item.Uninstaller.Location
+		// Run the installer
 		uninstallItem(item, itemURL, cachePath)
 	} else {
 		gorillalog.Warn("Unsupported item type", item.DisplayName, installerType)

--- a/pkg/installer/installer.go
+++ b/pkg/installer/installer.go
@@ -164,6 +164,12 @@ func uninstallItem(item catalog.Item, itemURL, cachePath string) string {
 	return uninstallerOut
 }
 
+var (
+	// By putting the functions in a variable, we can override later in tests
+	installItemFunc   = installItem
+	uninstallItemFunc = uninstallItem
+)
+
 // Install determines if action needs to be taken on a item and then
 // calls the appropriate function to install or uninstall
 func Install(item catalog.Item, installerType, urlPackages, cachePath string) string {
@@ -185,12 +191,12 @@ func Install(item catalog.Item, installerType, urlPackages, cachePath string) st
 		// Compile the item's URL
 		itemURL := urlPackages + item.Installer.Location
 		// Run the installer
-		installItem(item, itemURL, cachePath)
+		installItemFunc(item, itemURL, cachePath)
 	} else if installerType == "uninstall" {
 		// Compile the item's URL
 		itemURL := urlPackages + item.Uninstaller.Location
 		// Run the installer
-		uninstallItem(item, itemURL, cachePath)
+		uninstallItemFunc(item, itemURL, cachePath)
 	} else {
 		gorillalog.Warn("Unsupported item type", item.DisplayName, installerType)
 		return "Unsupported item type"

--- a/pkg/installer/installer_test.go
+++ b/pkg/installer/installer_test.go
@@ -190,7 +190,7 @@ func TestInstallItem(t *testing.T) {
 	// Check the result
 	nupkgCmd := filepath.Join(os.Getenv("ProgramData"), "chocolatey/bin/choco.exe")
 	nupkgFile := filepath.Join(pkgCache, nupkgPath)
-	expectedNupkg := "[" + nupkgCmd + " install " + nupkgFile + " -f -y -r]"
+	expectedNupkg := "[" + nupkgCmd + " install " + nupkgFile + " -f -y]"
 	if have, want := actualNupkg, expectedNupkg; have != want {
 		t.Errorf("\n-----\nhave\n%s\nwant\n%s\n-----", have, want)
 	}
@@ -311,7 +311,7 @@ func TestUninstallItem(t *testing.T) {
 	// Check the result
 	nupkgCmd := filepath.Join(os.Getenv("ProgramData"), "chocolatey/bin/choco.exe")
 	nupkgPath := filepath.Clean("testdata/packages/chef-client/chef-client-14.3.37-1-x64uninst.nupkg")
-	expectedNupkg := "[" + nupkgCmd + " uninstall " + nupkgPath + " -f -y -r]"
+	expectedNupkg := "[" + nupkgCmd + " uninstall " + nupkgPath + " -f -y]"
 	if have, want := actualNupkg, expectedNupkg; have != want {
 		t.Errorf("\n-----\nhave\n%s\nwant\n%s\n-----", have, want)
 	}

--- a/pkg/installer/installer_test.go
+++ b/pkg/installer/installer_test.go
@@ -22,6 +22,11 @@ var (
 	origExec            = execCommand
 	origCheckStatus     = statusCheckStatus
 	origReportInstalled = report.InstalledItems
+	origInstallItemFunc = installItemFunc
+
+	// These tore the URL that `Install` generates during testing
+	installItemURL   string
+	uninstallItemURL string
 
 	// Define a testing config for `download`
 	downloadCfg = config.Configuration{
@@ -463,4 +468,68 @@ func TestInstallReport(t *testing.T) {
 		t.Errorf("\nExpected: %#v\nReceived: %#v", expectedReport, report.InstalledItems)
 	}
 
+}
+
+func fakeInstallItem(item catalog.Item, itemURL, cachePath string) string {
+	installItemURL = itemURL
+	return ""
+}
+
+// TestInstallURL validates that the url for an installer is properly generated
+func TestInstallURL(t *testing.T) {
+	// Override checkStatus and installItemFunc with our fake versions
+	statusCheckStatus = fakeCheckStatus
+	installItemFunc = fakeInstallItem
+	defer func() {
+		statusCheckStatus = origCheckStatus
+		installItemFunc = origInstallItemFunc
+	}()
+
+	// Make sure the `installItemURL` variable is blank before we start
+	installItemURL = ""
+
+	// Run the msi installer with this status bypass checks
+	msiItem.DisplayName = statusActionNoError
+
+	// Run Install
+	Install(msiItem, "install", "https://example.com/", "testdata/")
+
+	// Check the result
+	expectedURL := "https://example.com/packages/chef-client/chef-client-14.3.37-1-x64.msi"
+
+	if have, want := installItemURL, expectedURL; have != want {
+		t.Errorf("\n-----\nhave\n%s\nwant\n%s\n-----", have, want)
+	}
+}
+
+func fakeUninstallItem(item catalog.Item, itemURL, cachePath string) string {
+	uninstallItemURL = itemURL
+	return ""
+}
+
+// TestUninstallURL validates that the url for an installer is properly generated
+func TestUninstallURL(t *testing.T) {
+	// Override checkStatus and installItemFunc with our fake versions
+	statusCheckStatus = fakeCheckStatus
+	uninstallItemFunc = fakeUninstallItem
+	defer func() {
+		statusCheckStatus = origCheckStatus
+		installItemFunc = origInstallItemFunc
+	}()
+
+	// Make sure the `installItemURL` variable is blank before we start
+	uninstallItemURL = ""
+
+	// Run the msi installer with this status bypass checks
+	msiItem.DisplayName = statusActionNoError
+
+	// Run Install
+	Install(msiItem, "uninstall", "https://example.com/", "testdata/")
+
+	// Check the result
+	expectedURL := "https://example.com/packages/chef-client/chef-client-14.3.37-1-x64.msi"
+
+	if have, want := installItemURL, expectedURL; have != want {
+		t.Errorf("\n-----\nhave\n%s\nwant\n%s\n-----", have, want)
+	}
 }

--- a/pkg/manifest/manifest.go
+++ b/pkg/manifest/manifest.go
@@ -1,7 +1,9 @@
 package manifest
 
 import (
+	"fmt"
 	"io/ioutil"
+	"os"
 	"path/filepath"
 
 	"github.com/1dustindavis/gorilla/pkg/config"
@@ -54,6 +56,14 @@ func Get(cfg config.Configuration) (manifests []Item, newCatalogs []string) {
 	// Add the top level manifest to the list
 	manifestsList = append(manifestsList, cfg.Manifest)
 
+	// Setup to catch a potential failure
+	defer func() {
+		if r := recover(); r != nil {
+			fmt.Println(r)
+			os.Exit(1)
+		}
+	}()
+
 	for manifestsRemaining > 0 {
 		currentManifest := manifestsList[manifestsProcessed]
 
@@ -65,7 +75,7 @@ func Get(cfg config.Configuration) (manifests []Item, newCatalogs []string) {
 		gorillalog.Info("Manifest Url:", manifestURL)
 		err := downloadFile(cfg.CachePath, manifestURL)
 		if err != nil {
-			gorillalog.Error("Unable to retrieve manifest:", currentManifest, err)
+			gorillalog.Error("Unable to retrieve manifest: ", err)
 		}
 
 		// Get new manifest

